File: financial-agent/client/index_nevermined.ts
/**
 * @fileoverview x402 client that implements the full payment flow:
 * 1. Makes initial request without token
 * 2. Receives 402 with PAYMENT-REQUIRED header
 * 3. Obtains x402 access token based on payment requirements
 * 4. Retries request with PAYMENT-SIGNATURE header
 */
import "dotenv/config";
import { Payments, EnvironmentName } from "@nevermined-io/payments";

// Configuration: Load environment variables with defaults
const AGENT_URL = process.env.AGENT_URL || "http://localhost:4001";
const SUBSCRIBER_API_KEY = process.env.SUBSCRIBER_NVM_API_KEY as string;
const NVM_ENVIRONMENT = (process.env.NVM_ENVIRONMENT ||
  "staging_sandbox") as EnvironmentName;

// Define demo conversation to show chatbot-style interaction
const DEMO_CONVERSATION_QUESTIONS = [
  "Hi there! I'm new to investing and keep hearing about diversification. Can you explain what that means in simple terms?",
  "That makes sense! So if I want to start investing but only have $100 a month, what should I focus on first?",
  "I've been thinking about cryptocurrency. What should a beginner like me know before investing in crypto?",
];

// Validate required environment variables
function validateEnvironment(): void {
  if (!SUBSCRIBER_API_KEY) {
    throw new Error("SUBSCRIBER_NVM_API_KEY is required in environment");
  }
}

// Initialize Nevermined Payments SDK
const payments = Payments.getInstance({
  nvmApiKey: SUBSCRIBER_API_KEY,
  environment: NVM_ENVIRONMENT,
});

// Interface for parsed payment requirements
interface PaymentRequired {
  x402Version: number;
  error?: string;
  resource: {
    url: string;
    description?: string;
    mimeType?: string;
  };
  accepts: Array<{
    scheme: string;
    network: string;
    planId: string;
    extra?: {
      version?: string;
      agentId?: string;
      httpVerb?: string;
    };
  }>;
  extensions: Record<string, unknown>;
}

// Parse the PAYMENT-REQUIRED header from a 402 response
function parsePaymentRequiredHeader(response: Response): PaymentRequired | null {
  const header = response.headers.get("payment-required");
  if (!header) {
    console.log("⚠️ No PAYMENT-REQUIRED header found in 402 response");
    return null;
  }

  try {
    const decoded = Buffer.from(header, "base64").toString("utf-8");
    const paymentRequired = JSON.parse(decoded) as PaymentRequired;
    console.log("📋 Received payment requirements:");
    console.log(`   Plan ID: ${paymentRequired.accepts[0]?.planId}`);
    console.log(`   Agent ID: ${paymentRequired.accepts[0]?.extra?.agentId}`);
    console.log(`   Scheme: ${paymentRequired.accepts[0]?.scheme}`);
    return paymentRequired;
  } catch (error) {
    console.error("Failed to parse PAYMENT-REQUIRED header:", error);
    return null;
  }
}

// Get x402 access token based on payment requirements
async function getX402AccessToken(planId: string, agentId: string): Promise<string> {
  console.log("🔐 Obtaining x402 access token...");

  // Try to get token directly first
  try {
    const result = await payments.x402.getX402AccessToken(planId, agentId);
    if (result?.accessToken) {
      console.log("✅ x402 access token obtained successfully");
      return result.accessToken;
    }
  } catch (tokenError: any) {
    console.log("⚠️ Could not get x402 token directly, checking subscription...");

    // Check if we need to purchase a subscription
    const balanceInfo: any = await payments.plans.getPlanBalance(planId);
    const hasCredits = Number(balanceInfo?.balance ?? 0) > 0;
    const isSubscriber = balanceInfo?.isSubscriber === true;

    if (!isSubscriber && !hasCredits) {
      console.log("💳 No subscription or credits found. Purchasing plan...");
      await payments.plans.orderPlan(planId);
    }

    // Retry getting the x402 token after subscription
    const result = await payments.x402.getX402AccessToken(planId, agentId);
    if (result?.accessToken) {
      console.log("✅ x402 access token obtained successfully");
      return result.accessToken;
    }
  }

  throw new Error("Failed to obtain x402 access token");
}

// Simple loading animation for terminal
function startLoadingAnimation(): () => void {
  const frames = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];
  let i = 0;
  const interval = setInterval(() => {
    process.stdout.write(`\r${frames[i]} FinGuide is thinking...`);
    i = (i + 1) % frames.length;
  }, 100);

  return () => {
    clearInterval(interval);
    process.stdout.write("\r");
  };
}

// Cache for x402 tokens (keyed by planId:agentId)
const tokenCache = new Map<string, string>();

// Send a question to the protected financial agent using x402 flow
async function askAgent(
  input: string,
  sessionId?: string
): Promise<{ output: string; sessionId: string; payment?: any }> {
  // Start loading animation
  const stopLoading = startLoadingAnimation();

  try {
    // Prepare request payload
    const requestBody = {
      input_query: input,
      sessionId: sessionId,
    };

    // Step 1: Make initial request without PAYMENT-SIGNATURE header
    console.log("\n📤 Making request to agent...");
    const initialResponse = await fetch(`${AGENT_URL}/ask`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(requestBody),
    });

    // Step 2: If we get 402, parse PAYMENT-REQUIRED header
    if (initialResponse.status === 402) {
      console.log("💳 Received 402 Payment Required");

      const paymentRequired = parsePaymentRequiredHeader(initialResponse);
      if (!paymentRequired || !paymentRequired.accepts[0]) {
        throw new Error("Invalid payment requirements in 402 response");
      }

      const scheme = paymentRequired.accepts[0];
      const planId = scheme.planId;
      const agentId = scheme.extra?.agentId;

      if (!planId || !agentId) {
        throw new Error("Missing planId or agentId in payment requirements");
      }

      // Step 3: Get x402 access token (use cache if available)
      const cacheKey = `${planId}:${agentId}`;
      let x402Token = tokenCache.get(cacheKey);

      if (!x402Token) {
        x402Token = await getX402AccessToken(planId, agentId);
        tokenCache.set(cacheKey, x402Token);
      } else {
        console.log("🔑 Using cached x402 access token");
      }

      // Step 4: Retry request with PAYMENT-SIGNATURE header (x402 standard)
      console.log("📤 Retrying request with PAYMENT-SIGNATURE header...");
      const retryResponse = await fetch(`${AGENT_URL}/ask`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "PAYMENT-SIGNATURE": x402Token, // x402 token is already base64 encoded
        },
        body: JSON.stringify(requestBody),
      });

      // Handle retry response
      if (!retryResponse.ok) {
        // If still 402, token might be invalid - clear cache and throw
        if (retryResponse.status === 402) {
          tokenCache.delete(cacheKey);
          const errorBody = await retryResponse.text().catch(() => "");
          throw new Error(`Payment still required after retry: ${errorBody}`);
        }
        throw new Error(
          `Agent request failed: ${retryResponse.status} ${retryResponse.statusText}`
        );
      }

      // Parse PAYMENT-RESPONSE header if present
      const paymentResponseHeader = retryResponse.headers.get("payment-response");
      if (paymentResponseHeader) {
        try {
          const paymentResponse = JSON.parse(
            Buffer.from(paymentResponseHeader, "base64").toString("utf-8")
          );
          console.log("✅ Payment settled:", paymentResponse.success ? "Success" : "Failed");
          if (paymentResponse.transaction) {
            console.log(`   Transaction: ${paymentResponse.transaction.substring(0, 20)}...`);
          }
        } catch {
          // Ignore parsing errors for payment response header
        }
      }

      return (await retryResponse.json()) as {
        output: string;
        sessionId: string;
        payment?: any;
      };
    }

    // If not 402, handle as normal response or error
    if (!initialResponse.ok) {
      throw new Error(
        `Agent request failed: ${initialResponse.status} ${initialResponse.statusText}`
      );
    }

    return (await initialResponse.json()) as {
      output: string;
      sessionId: string;
      payment?: any;
    };
  } finally {
    // Stop loading animation
    stopLoading();
  }
}

/**
 * Run the x402 demo client.
 * Demonstrates the full x402 payment flow with PAYMENT-REQUIRED and PAYMENT-SIGNATURE headers.
 */
async function runDemo(): Promise<void> {
  console.log("🚀 Starting Financial Agent Demo (x402 Payment Flow)\n");
  console.log("This demo implements the full x402 payment protocol:");
  console.log("  1. Client makes request without payment");
  console.log("  2. Server returns 402 with PAYMENT-REQUIRED header");
  console.log("  3. Client obtains x402 access token");
  console.log("  4. Client retries with PAYMENT-SIGNATURE header");
  console.log("  5. Server verifies, executes, settles, returns PAYMENT-RESPONSE\n");

  // Validate environment setup
  validateEnvironment();

  // Track session across multiple questions
  let sessionId: string | undefined;

  // Send each demo question using x402 flow
  for (let i = 0; i < DEMO_CONVERSATION_QUESTIONS.length; i++) {
    const question = DEMO_CONVERSATION_QUESTIONS[i];

    console.log("=".repeat(80));
    console.log(`📝 Question ${i + 1}: ${question}`);

    try {
      // Send question using x402 payment flow
      const result = await askAgent(question, sessionId);

      // Update sessionId for next question
      sessionId = result.sessionId;

      // Display agent response
      console.log(`\n🤖 FinGuide (Session: ${sessionId}):`);
      console.log(result.output);

      // Display payment info if available
      if (result.payment) {
        console.log(`\n💰 Payment Info:`);
        console.log(`   Credits used: ${result.payment.creditsRedeemed || "unknown"}`);
        console.log(`   Remaining balance: ${result.payment.remainingBalance || "unknown"}`);
      }

      console.log("\n");
    } catch (error) {
      console.error(`\n❌ Error processing question ${i + 1}:`, error);
      break;
    }
  }

  console.log("=".repeat(80));
  console.log("✅ Demo completed!");
}

// Run the demo and handle any errors
runDemo().catch((error) => {
  console.error("💥 Demo failed:", error);
  process.exit(1);
});

File: financial-agent-py/client/index_nevermined.py
"""
Minimal client that sends three questions to the agent,
storing and reusing the returned sessionId to maintain conversation context.
"""
import os
import sys
import asyncio
import httpx
from typing import Dict, Any, Optional
from dotenv import load_dotenv
from payments_py import Payments, PaymentOptions

# Load environment variables
load_dotenv()

# Configuration: Load environment variables with defaults
AGENT_URL = os.getenv("AGENT_URL", "http://localhost:8000")
PLAN_ID = os.getenv("NVM_PLAN_ID", "")
AGENT_ID = os.getenv("NVM_AGENT_ID", "")
SUBSCRIBER_API_KEY = os.getenv("SUBSCRIBER_NVM_API_KEY", "")
NVM_ENVIRONMENT = os.getenv("NVM_ENVIRONMENT", "sandbox")

# Define demo conversation to show chatbot-style interaction
DEMO_CONVERSATION_QUESTIONS = [
    "Hi there! I'm new to investing and keep hearing about diversification. Can you explain what that means in simple terms?",
    "That makes sense! So if I want to start investing but only have $100 a month, what should I focus on first?",
    "I've been thinking about cryptocurrency. What should a beginner like me know before investing in crypto?",
]

# Simple loading animation frames
LOADING_FRAMES = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏']

class LoadingAnimation:
    def __init__(self, message: str = "FinGuide is thinking..."):
        self.message = message
        self.running = False
        self.frame_index = 0

    async def start(self):
        self.running = True
        while self.running:
            frame = LOADING_FRAMES[self.frame_index % len(LOADING_FRAMES)]
            print(f"\r{frame} {self.message}", end="", flush=True)
            self.frame_index += 1
            await asyncio.sleep(0.1)

    def stop(self):
        self.running = False
        print("\r", end="", flush=True)  # Clear the line

def validate_environment() -> None:
    """Validate required environment variables"""
    if not PLAN_ID or not AGENT_ID:
        raise Exception("NVM_PLAN_ID and NVM_AGENT_ID are required in environment")
    if not SUBSCRIBER_API_KEY:
        raise Exception("SUBSCRIBER_NVM_API_KEY is required in environment")

async def get_access_token() -> str:
    """Get or purchase access token for protected agent"""
    print("🔐 Setting up Nevermined access...")

    # Initialize Nevermined Payments SDK
    payments = Payments.get_instance(PaymentOptions(
        nvm_api_key=SUBSCRIBER_API_KEY,
        environment=NVM_ENVIRONMENT,
    ))

    # Check current plan balance and subscription status
    # NOTE: get_plan_balance raises PaymentsError if the plan doesn't exist or if there's an API error
    # The returned balance_info is a validated Pydantic object with proper types
    balance_info = payments.plans.get_plan_balance(PLAN_ID)

    # Access balance and subscription status directly from the Pydantic object
    has_credits = balance_info.balance > 0
    is_subscriber = balance_info.is_subscriber

    # Purchase plan if not subscribed and no credits
    if not is_subscriber and not has_credits:
        print("💳 No subscription or credits found. Purchasing plan...")
        payments.plans.order_plan(PLAN_ID)

    # Get access token for the agent
    credentials = payments.agents.get_agent_access_token(PLAN_ID, AGENT_ID)

    if not credentials or not credentials.get("accessToken"):
        raise Exception("Failed to obtain access token")

    print("✅ Access token obtained successfully")
    return credentials["accessToken"]

async def ask_agent(input_text: str, access_token: str, session_id: Optional[str] = None) -> Dict[str, Any]:
    """Send a question to the protected financial agent"""

    # Start loading animation
    loading = LoadingAnimation()
    loading_task = asyncio.create_task(loading.start())

    try:
        # Prepare request payload
        request_body = {
            "input_query": input_text,
            "sessionId": session_id
        }

        # Prepare headers with authorization
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {access_token}"
        }

        async with httpx.AsyncClient() as client:
            # Make HTTP request to protected agent
            response = await client.post(
                f"{AGENT_URL}/ask",
                json=request_body,
                headers=headers,
                timeout=30.0
            )

            # Handle HTTP errors (including payment required)
            if response.status_code == 402:
                raise Exception("Payment Required - insufficient credits or subscription")
            elif response.status_code != 200:
                error_text = response.text
                raise Exception(f"Agent request failed: {response.status_code} {response.reason_phrase} {error_text}")

            # Parse and return JSON response
            return response.json()

    finally:
        # Stop loading animation
        loading.stop()
        loading_task.cancel()
        try:
            await loading_task
        except asyncio.CancelledError:
            pass

async def run_demo() -> None:
    """
    Run the protected demo client.
    Sends predefined financial questions to the agent with Authorization and reuses sessionId to preserve context.
    """
    print("🚀 Starting Financial Agent Demo (Protected with Nevermined)\n")

    # Validate environment setup
    validate_environment()

    # Obtain access token for protected agent
    access_token = await get_access_token()

    # Track session across multiple questions
    session_id: Optional[str] = None

    # Send each demo question and maintain conversation context
    for i, question in enumerate(DEMO_CONVERSATION_QUESTIONS):
        print(f"📝 Question {i + 1}: {question}")

        try:
            # Send question to protected agent (reusing sessionId for context)
            result = await ask_agent(question, access_token, session_id)

            # Update sessionId for next question
            session_id = result["sessionId"]

            # Display agent response and payment info
            print(f"🤖 FinGuide (Session: {session_id}):")
            print(result["output"])

            if result.get("redemptionResult"):
                credits_redeemed = result["redemptionResult"].get("creditsRedeemed", 0)
                print(f"💰 Credits redeemed: {credits_redeemed}")

            print("\n" + "=" * 80 + "\n")

        except Exception as error:
            print(f"❌ Error processing question {i + 1}: {error}")
            break

    print("✅ Demo completed!")

# Run the demo and handle any errors
async def main():
    try:
        await run_demo()
    except Exception as error:
        print(f"💥 Demo failed: {error}")
        sys.exit(1)

if __name__ == "__main__":
    asyncio.run(main())